File: client/src/store/domain/Character.js
import IdRequiredException from './IdRequiredException';

const NO_ID = null;
const DEFAULT_NAME = 'No name character';
export const DEFAULT_REQUIRED_SLEEP_MINUTES = 6 * 60;

export default class Character {
    constructor({id = NO_ID, name = DEFAULT_NAME, requiredSleepTime = DEFAULT_REQUIRED_SLEEP_MINUTES}){
        if(id === NO_ID){
            throw new IdRequiredException();
        }
        this._id = id;
        this._name = name;
        this._requiredSleepTime = requiredSleepTime;
    }

    get id(){
        return this._id;
    }

    get name() {
        return this._name;
    }

    get requiredSleepTime() {
        return this._requiredSleepTime;
    }

    copy = () => new Character(this);
}

File: optimizer/src/main/java/es/carloscasalar/nwp/domain/Character.java
package es.carloscasalar.nwp.domain;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;

import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.ArrayList;
import java.util.List;

/**
 * Character
 */
@AllArgsConstructor
@Data
@Builder
@EqualsAndHashCode(exclude = {"senses", "requiredSleepTime"})
@ToString(exclude = {"senses", "requiredSleepTime"})
public class Character {
    @JsonProperty("name")
    @NotNull
    private final String name;

    @JsonProperty("senses")
    @NotNull
    @Size(min = 1)
    private final List<Sense> senses;

    @JsonProperty("requiredSleepTime")
    @Min(0)
    private final Integer requiredSleepTime;

    public Character copy() {
        return Character.builder()
                .name(name)
                .senses(new ArrayList<>(senses))
                .requiredSleepTime(requiredSleepTime)
                .build();
    }

    public int overSleepTime(int totalTimeSleeping) {
        int excess = totalTimeSleeping - requiredSleepTime;
        return Math.max(excess, 0);
    }

    public boolean needsToSleep() {
        return requiredSleepTime > 0;
    }
}
